actions/service.go

```diff
@@ -102,40 +121,199 @@ func (s *Service) loadMatchedActions(ctx context.Context, source Source, spec Ma
 func (s *Service) allocateTasks(runID string, actions []*Action) ([]*Task, error) {
 	var tasks []*Task
 	for _, action := range actions {
+		var prevTask *Task
 		for _, hook := range action.Hooks {
 			h, err := NewHook(hook, action)
 			if err != nil {
 				return nil, err
 			}
-			tasks = append(tasks, &Task{
-				RunID:  runID,
-				Action: action,
-				HookID: hook.ID,
-				Hook:   h,
-			})
+			task := &Task{
+				RunID:     runID,
+				HookRunID: graveler.NewRunID(),
+				Action:    action,
+				HookID:    hook.ID,
+				Hook:      h,
+			}
+			// append new task or chain to the last one based on the current action
+			if prevTask == nil {
+				tasks = append(tasks, task)
+			} else {
+				prevTask.NextTask = task
+			}
+			prevTask = task
 		}
 	}
 	return tasks, nil
 }
 
-func (s *Service) runTasks(ctx context.Context, hooks []*Task, event Event, deps Deps) error {
+func (s *Service) runTasks(ctx context.Context, record graveler.HookRecord, tasks []*Task) error {
 	var g multierror.Group
-	for _, h := range hooks {
-		hh := h // pinning
+	for _, task := range tasks {
+		task := task // pin
 		g.Go(func() error {
-			return hh.Hook.Run(ctx, event, &HookOutputWriter{
-				RunID:      hh.RunID,
-				ActionName: hh.Action.Name,
-				HookID:     hh.HookID,
-				Writer:     deps.Output,
-			})
+			for task != nil {
+				hookOutputWriter := &HookOutputWriter{
+					Writer:           s.Writer,
+					StorageNamespace: record.StorageNamespace.String(),
+					RunID:            task.RunID,
+					HookRunID:        task.HookRunID,
+					ActionName:       task.Action.Name,
+					HookID:           task.HookID,
+				}
+				task.StartTime = time.Now()
+				task.Err = task.Hook.Run(ctx, record, hookOutputWriter)
+				task.EndTime = time.Now()
+
+				if task.Err != nil {
+					// wrap error with more information and return
+					task.Err = fmt.Errorf("run '%s' failed on action '%s' hook '%s': %w",
+						task.RunID, task.Action.Name, task.HookID, task.Err)
+					return task.Err
+				}
+
+				task = task.NextTask
+			}
+			return nil
 		})
 	}
 	return g.Wait().ErrorOrNil()
 }
 
-func NewRunID(t time.Time) string {
-	uid := strings.ReplaceAll(uuid.New().String(), "-", "")
-	runID := t.UTC().Format(time.RFC3339) + "_" + uid
-	return runID
+func (s *Service) insertRunInformation(ctx context.Context, record graveler.HookRecord, tasks []*Task) error {
```

Comment:
Name is a little bit misleading since it will only work for Runs which are `pre_*` (since commitID isn't set for them).
Maybe just mention that in the func comment?

Reply:
Updated the code to update 'commit_id' - currently, this code updates information just for preXXX operations, but it supports update information for postXXX events too